File: app/src/main/java/com/github/aldychris/mobiusexample/CounterLogic.java
package com.github.aldychris.mobiusexample;

import com.spotify.mobius.Effects;
import com.spotify.mobius.Next;


class CounterLogic {
    static int update(int counter, CounterEvent counterEvent) {
        switch (counterEvent) {
            case UP:
                return counter + 1;

            case DOWN:
                if (counter > 0) {
                    return counter - 1;
                }
            default: return counter;
        }
    }

    static Next<Integer, CounterEffect> updateWithEffect(int counter, CounterEvent counterEvent){
        switch (counterEvent) {
            case UP:
                return Next.next(counter + 1);
            case DOWN:
                if (counter > 0)
                    return Next.next(counter - 1);

                return Next.next(counter, Effects.effects(CounterEffect.REPORT_ERROR_NEGATIVE));

            default: return Next.next(counter);
        }
    }

}

File: app/src/main/java/com/github/aldychris/mobiusexample/CounterLogic.kt
package com.github.aldychris.mobiusexample

import com.spotify.mobius.Effects
import com.spotify.mobius.Next

internal object CounterLogic {

    @JvmStatic
    fun updateWithEffect(counter: Int, counterEvent: CounterEvent?): Next<Int, CounterEffect> {
        return when (counterEvent) {
            Up -> Next.next(counter + 1)
            Down -> {
                return if (counter > 0) {
                    Next.next(counter - 1)
                } else {
                    Next.next(counter,Effects.effects(IsError))
                }
            }
            else -> Next.next(counter)
        }
    }

}